### DISCUSSION TITLE
CompletionProvider for Monaco

### DISCUSSION THREAD

**Discussion by Lemour-sudo** [eclipse-langium/langium#1454] [Q&A]
Is there a way to get the `CompletionItemProvider` for a language from Langium such that it can be used to register a completion provider for a language in [monaco](https://github.com/microsoft/vscode/blob/258537d40b084b8b51c6992ac5a75a1421db3219/src/vs/editor/standalone/browser/standaloneLanguages.ts#L613)

The lsp is a great solution in most cases, but it easily results in large bundle sizes which can not be deployed on sites running on limited memory environments. Having provider object for functionalities like completion and diagnostics that can be registered onto monaco will be of great help.

**Comment by msujew** [comment]
Hey @Lemour-sudo,

you can always just use the `CompletionProvider` of the langium service container object and populate the monaco completion provider that way. Note that this requires manual document handling, since the `CompletionProvider` expects a `LangiumDocument` to be available to perform completion on top of.

On a performance note: This could easily block the main thread of monaco, since the completion computation (and whole document handling) can be pretty expensive on large documents. So you should attempt to do this in a worker thread.

> The lsp is a great solution in most cases, but it easily results in large bundle sizes which can not be deployed on sites running on limited memory environments.

I'm a bit confused by that. Langium in itself is roughly ~300kb minified. Adding a language on top of that is something in the range of ~200kb + any additional code that you have added to the language implementation. Using Langium in a language server in a separate bundle or in the main bundle for monaco shouldn't have any influence on total build size. You will always need to add the Langium dependency (+ transitive dependencies) and your language code to the bundle. Or is it the monaco-languageclient support that is actually the heavy stuff?

**Comment by Lemour-sudo** [comment]
I will look more into the provider required to register in monaco. I was under the impression that this would be possible without requiring a `LangiumDocument` to be passed.



> Or is it the monaco-languageclient support that is actually the heavy stuff?

Yes. I have been trying to use the Langium lsp through the monaco-languageclient and the size of the bundle increases drastically to about 9mb. I added an issue on their repo to query on this problem: https://github.com/CodinGame/monaco-vscode-api/issues/383. It turns out that it is inevitable to initialize the required vscode services in order to use the client that can be used to communicate with the server.

**Reply by msujew** [reply]
Langium is written in a way that always assumes that for any given operation, a `LangiumDocument` is available to operate on. In fact, I don't think there's any other way to implement this feature.

Creating a document manually should be fairly straightforward using the `TextDocumentFactory` service.
